### ISSUE TITLE
Error while increasing h2o cluster size > 2 on H2O Kubernetes external sparkling water setup

### ISSUE THREAD

**Issue by Abhishekmamidi123** [h2oai/sparkling-water#2538]
I'm using H2OXGBoostClassifier to build a model on spark data. When I try to increase spark.ext.h2o.external.cluster.size more than 2, the h2o cluster doesn't start and I get the below error:

```
Traceback (most recent call last):
  File "/tmp/spark-f95316f7-3eb2-4e5f-a7be-13f0a92abc90/xclassifier_testing.py", line 11, in <module>
    hc = H2OContext.getOrCreate()
  File "/opt/conda/lib/python3.8/site-packages/ai/h2o/sparkling/H2OContext.py", line 89, in getOrCreate
    jhc = module.getOrCreate(selected_conf._jconf)
  File "/opt/spark/python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py", line 1304, in __call__
  File "/opt/spark/python/lib/pyspark.zip/pyspark/sql/utils.py", line 111, in deco
  File "/opt/spark/python/lib/py4j-0.10.9-src.zip/py4j/protocol.py", line 326, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling o82.getOrCreate.
: java.lang.RuntimeException: Timeout during clouding of external H2O backend on K8s.
```

I tried increasing spark.ext.h2o.cluster.start.timeout to 240s and conf spark.ext.h2o.cloud.timeout to 180000 ms, but it didn't work. The main goal here is to increase the cluster size so that the H2O has more memory to train the model. Please let me know if my understanding is wrong in terms of the Kubernetes setup. Will it automatically increase based on the requirement? Or does it make sense to start more h2o clusters?

I also made sure that the spark.executors is the same as the cluster.size, but it didn't help.

Below is the spark submit command I'm using:

```
spark-submit \
--master "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" \
--deploy-mode cluster \
--conf spark.scheduler.minRegisteredResourcesRatio=1 \
--conf spark.kubernetes.container.image='file-path/spark-churn:h2o-3.1' \
--conf spark.kubernetes.container.image.pullPolicy=IfNotPresent \
--conf spark.kubernetes.container.image.pullSecrets='xxxxxxxxxxxxxx' \
--conf spark.ext.h2o.backend.cluster.mode=external \
--conf spark.ext.h2o.external.start.mode=auto \
--conf spark.ext.h2o.external.auto.start.backend=kubernetes \
--conf spark.ext.h2o.external.cluster.size=4 \
--conf spark.ext.h2o.external.memory=8G \
--conf spark.ext.h2o.external.k8s.docker.image=h2oai/sparkling-water-external-backend:3.32.1.3-1-3.1 \
--conf spark.kubernetes.authenticate.driver.serviceAccountName=xxx-spark \
--conf spark.kubernetes.namespace=xxx-ds \
--conf spark.eventLog.enabled=true \
--conf spark.eventLog.dir=s3a://file-path/spark-history \
--conf spark.eventLog.logStageExecutorMetrics=true \
--conf spark.dynamicAllocation.enabled=true \
--conf spark.dynamicAllocation.shuffleTracking.enabled=true \
--conf spark.dynamicAllocation.executorIdleTimeout=43200s \
--conf spark.kubernetes.submission.waitAppCompletion=true \
--conf spark.kubernetes.node.selector.nodegroup=sparknodes2 \
--name h2o-testing \
--executor-memory 8g \
--driver-memory 8g \
--executor-cores 2 \
--driver-cores 2 \
s3a://file_path/xclassifier_testing.py 
```

I am sharing the code for reference. I am actually doing GridSearch, but the below code is also giving the same issue, as the issue is not related to code, but more of a configuration issue:

```
from pysparkling.ml import H2OGridSearch, H2OXGBoostRegressor, H2OXGBoost, H2OMOJOModel, H2OXGBoostClassifier
model_kwargs = {'labelCol': 'target_variable',
                'nfolds': 0,
                'splitRatio': 0.8,
                'ignoredCols': ["customer_id"],
                'columnsToCategorical': [], 
                'ignoreConstCols': True,
                'predictionCol': "prediction",
                'withContributions': False}
estimator = H2OXGBoostClassifier(**model_kwargs)
XGBoost_model = estimator.fit(data)
predictions = XGBoost_model.transform(data)
predictions.show(5, truncate=False)
```

In addition to this, what are parameters needs to be adjusted so that the "H2OXGBoostClassifier + GridSearch" scales on H2O Kubernetes sparkling water setup for large data?

To give more context behind the issue:
I am training the XGBoost model with the above setup on large data ~1.4GB. It enters into modelling and I see the below logs. The below logs continues until the modeling is completed. When I ran on ~700 MB (10M records), it took 170 minutes and completed successfully. For ~1.4 GB (20M records), it's not coming out. I waited for more than 12 hrs and it still didn't complete training.

```
INFO H2OJob: H2O node http://h2o-service.default.svc.cluster.local:54321/3/Jobs/$0301ac1f009932d4ffffffff$_9a9f279b290a616a87b345c5533f45df successfully responded for the GET.
INFO H2OJob: Waiting for job $0301ac1f009932d4ffffffff$_9a9f279b290a616a87b345c5533f45df to finish...
21/06/22 14:55:26 INFO H2OJob: H2O node http://h2o-service.default.svc.cluster.local:54321/3/Jobs/$0301ac1f009932d4ffffffff$_9a9f279b290a616a87b345c5533f45df successfully responded for the GET.
INFO H2OJob: Waiting for job $0301ac1f009932d4ffffffff$_9a9f279b290a616a87b345c5533f45df to finish...
INFO H2OJob: H2O node http://h2o-service.default.svc.cluster.local:54321/3/Jobs/$0301ac1f009932d4ffffffff$_9a9f279b290a616a87b345c5533f45df successfully responded for the GET.
```

In summary, I am trying to increase the training speed. So, what are the parameters that can be changed? As I was trying different parameters, I tried increasing spark executors and h2o cluster size and I faced this issue. Please help me here.

Also, if you can give inputs on the best values to choose for the below parameters to decrease training time. I understand that it depends on the training data size, but I want to understand in relative terms. For example, do we need to make sure that cluster size and executors same? etc.
--executor-memory 8g \
--driver-memory 8g \
--executor-cores 2 \
--driver-cores 2 \
--conf spark.executor.instances=2 \
--conf spark.ext.h2o.external.cluster.size=4 \
--conf spark.ext.h2o.external.memory=8G \

Should I also add memoryOverhead for the Kubernetes setup? If yes, what is parameter should I use? I read that it should be 120% than spark executor memory. But, I didn't find any config param for kubernetes.

I am getting this warning sometimes "root failed". Can I ignore this? or will it cause any issues?
```
INFO H2OContext: Trying to lock H2O cluster h2o-service.default.svc.cluster.local:54321 - root.
21/06/22 17:43:32 WARN H2OContext: Locking of the H2O cluster h2o-service.default.svc.cluster.local:54321 - root failed.
ai.h2o.sparkling.backend.exceptions.RestApiNotReachableException: H2O node http://h2o-service.default.svc.cluster.local:54321 is not reachable.
Please verify that you are passing ip and port of existing cluster node and the cluster
is running with web enabled.
```

Please share any documentation, if any.

I really appreciate it if you can help me in solving the issue. Thanks in advance :)

@jakubhava: Please provide some inputs on this issue

**Comment by mn-mikke** [comment]
@Abhishekmamidi123 At first look, I don't see anything wrong in your configuration.  Good starting point is a configuration in [this tutorial](https://docs.h2o.ai/sparkling-water/3.1/latest-stable/doc/deployment/kubernetes.html#automatic-mode-of-external-backend). it's covered by automated tests.

Can check if pods with H2O nodes were created? Did your k8s cluster have enough resources?
hint: ```kubectl get events --sort-by='.lastTimestamp'```

If the given number of H2O node pods was created, can you share logs from the pods + logs from h2o driver?


> Should I also add memoryOverhead for the Kubernetes setup? If yes, what is parameter should I use? I read that it should be 120% than spark executor memory. But, I didn't find any config param for kubernetes.

If you use SW internal backend, yes. And you can follow [this documentation](https://docs.h2o.ai/sparkling-water/3.1/latest-stable/doc/ml/sw_xgboost.html#xgboost-memory-configuration). But in your case of the external backend and docker image you don't have to worry about that. `h2oai/sparkling-water-external-backend:3.32.1.3-1-3.1`. The image starts H2O node with  the java option `-XX:MaxRAMPercentage=50`, which gives XGBoost 50% of native memory.

**Comment by Abhishekmamidi123** [comment]
Thank you so much for the confirmation on memory overhead wrt SW external backend. I have been referring to the above documentation pages that you shared for setting up and using.

I have few queries:

1. k8s cluster has enough resources. I don't see any pods with H2O nodes. Btw, how can we check that? I can see a lot of executor pods. I used the above command that you shared.
2. For example, I have 3GB of data and the h2o external memory is 4GB, can the Kubernetes external mode scale out of the box depending on the memory requirement. If not, how much GB should I specify based on the input data size OR what are other configurations needs to be made? Any inputs on this.
3. When should we increase spark executor memory and when to increase h2o.external.memory? Should we keep it the same for consistency?

I tried increasing cores, memory, etc. but, nothing worked on 1.4GB data (~20M samples) when training using H2OXGBoost GridSearch. It keeps failing. I want to try increasing cluster size, but it's not taking values greater than 2.

@mn-mikke @jakubhava: Please share your thoughts on the above comments. It would be very helpful for our experiments. Thanks in advance :)

**Comment by Madaditya** [comment]
Hi @mn-mikke ,

I am working with @Abhishekmamidi123 on the h2o issue. 

**kubectl get events** shows that pods are getting assigned and created.


`2m37s       Warning   FailedToUpdateEndpointSlices   service/ak-h2o-mra-100k-exp5-2cd0777a395b5322-driver-svc   Error updating Endpoint Slices for Service ua-ds/ak-h2o-mra-100k-exp5-2cd0777a395b5322-driver-svc: failed to update ak-h2o-mra-100k-exp5-2cd0777a395b5322-driver-svc-cmrw4 EndpointSlice for Service ua-ds/ak-h2o-mra-100k-exp5-2cd0777a395b5322-driver-svc: endpointslices.discovery.k8s.io "ak-h2o-mra-100k-exp5-2cd0777a395b5322-driver-svc-cmrw4" not found
2m32s       Warning   FailedMount                    pod/ak-h2o-mra-100k-exp5-1013c07a395d1418-driver           MountVolume.SetUp failed for volume "spark-conf-volume-driver" : configmap "spark-drv-84056f7a395d16c3-conf-map" not found
2m32s       Normal    Scheduled                      pod/ak-h2o-mra-100k-exp5-1013c07a395d1418-driver           Successfully assigned ua-ds/ak-h2o-mra-100k-exp5-1013c07a395d1418-driver to ip-172-31-1-123.ec2.internal
2m31s       Normal    Pulled                         pod/ak-h2o-mra-100k-exp5-1013c07a395d1418-driver           Container image "personalizeai.azurecr.io/ua/spark-churn:h2o-3.1" already present on machine
2m31s       Normal    Created                        pod/ak-h2o-mra-100k-exp5-1013c07a395d1418-driver           Created container spark-kubernetes-driver
2m30s       Normal    Started                        pod/ak-h2o-mra-100k-exp5-1013c07a395d1418-driver           Started container spark-kubernetes-driver
44s         Normal    Scheduled                      pod/ak-h2o-mra-100k-exp5-d2064f7a395d30c6-exec-1           Successfully assigned ua-ds/ak-h2o-mra-100k-exp5-d2064f7a395d30c6-exec-1 to ip-172-31-0-120.ec2.internal
43s         Normal    Created                        pod/ak-h2o-mra-100k-exp5-d2064f7a395d30c6-exec-1           Created container spark-kubernetes-executor
43s         Normal    Started                        pod/ak-h2o-mra-100k-exp5-d2064f7a395d30c6-exec-1           Started container spark-kubernetes-executor
43s         Normal    Pulled                         pod/ak-h2o-mra-100k-exp5-d2064f7a395d30c6-exec-1           Container image "personalizeai.azurecr.io/ua/spark-churn:h2o-3.1" already present on machine
23s         Normal    Scheduled                      pod/ak-h2o-mra-100k-exp5-d2064f7a395d30c6-exec-2           Successfully assigned ua-ds/ak-h2o-mra-100k-exp5-d2064f7a395d30c6-exec-2 to ip-172-31-0-120.ec2.internal
21s         Normal    Started                        pod/ak-h2o-mra-100k-exp5-d2064f7a395d30c6-exec-2           Started container spark-kubernetes-executor
21s         Normal    Created                        pod/ak-h2o-mra-100k-exp5-d2064f7a395d30c6-exec-2           Created container spark-kubernetes-executor
21s         Normal    Pulled                         pod/ak-h2o-mra-100k-exp5-d2064f7a395d30c6-exec-2           Container image "personalizeai.azurecr.io/ua/spark-churn:h2o-3.1" already present on machine`

Also, a lot of hardware is available on the EKS cluster

![image](https://user-images.githubusercontent.com/2889089/123120129-b34cc080-d461-11eb-9b4e-ca54a93771e8.png)

**Comment by mn-mikke** [comment]
Great. 

@Madaditya @Abhishekmamidi123 Can you share logs from H2O node pods and Spark driver pod?

**Comment by Abhishekmamidi123** [comment]
[logs-from-spark-kubernetes-driver-in-h2o-test2-mra-20m-9ae7017a1dcff367-driver - cloud timeout h2o.txt](https://github.com/h2oai/sparkling-water/files/6703467/logs-from-spark-kubernetes-driver-in-h2o-test2-mra-20m-9ae7017a1dcff367-driver.-.cloud.timeout.h2o.txt)


@mn-mikke: I have attached the file. As I said, it's stopping while creating the H2O cluster when h2o.cluster.size is greater than 2.

Can you please give some inputs on how to get the logs of H2O node pods? We couldn't locate H2O pods.

Can you also give some inputs on the above queries I posted? Thanks in advance

**Comment by Abhishekmamidi123** [comment]
@mn-mikke @jakubhava: Can you please guide us here? Your inputs would be really helpful.

**Comment by Abhishekmamidi123** [comment]
@mn-mikke @jakubhava: Here are more clear experiment analysis and results. This should give you more insights into the issue.

We are using m5.4xlarge which is (16 core, 64 GB ram). This can scale up to 10 nodes. Below is the spark config:

```
spark-submit \
--master "k8s://XXXXXXXXXXXXXXXXXXxxxx.yy7.us-east-1.eks.amazonaws.com:443" \
--deploy-mode cluster \
--conf spark.scheduler.minRegisteredResourcesRatio=1 \
--conf spark.kubernetes.container.image='spark-churn:h2o-3.1' \
--conf spark.kubernetes.container.image.pullPolicy=IfNotPresent \
--conf spark.kubernetes.container.image.pullSecrets='personalizeai-repo-secret' \
--conf spark.executor.instances=2 \
--conf spark.ext.h2o.backend.cluster.mode=external \
--conf spark.ext.h2o.external.start.mode=auto \
--conf spark.ext.h2o.external.auto.start.backend=kubernetes \
--conf spark.ext.h2o.external.cluster.size=2 \
--conf spark.ext.h2o.external.memory=28G \
--conf spark.ext.h2o.external.k8s.docker.image=h2oai/sparkling-water-external-backend:3.32.1.3-1-3.1 \
--conf spark.eventLog.enabled=true \
--conf spark.eventLog.dir=s3a://file-path/spark-history \
--conf spark.eventLog.logStageExecutorMetrics=true \
--conf spark.dynamicAllocation.enabled=true \
--conf spark.dynamicAllocation.shuffleTracking.enabled=true \
--conf spark.dynamicAllocation.executorIdleTimeout=43200s \
--conf spark.kubernetes.submission.waitAppCompletion=true \
--conf spark.kubernetes.node.selector.nodegroup=sparknodes \
--name h2o-mra-30M-exp6 \
--executor-memory 12g \
--driver-memory 8g \
--executor-cores 2 \
--driver-cores 2 \
s3a://file-path/temp/xclassifier_testing.py
```
If you can see above, the h2o cluster size is 2 and external memory is 28 GB. Below is the H2O cluster config logs. I have attached the complete logs as well. The logs show that H2O cluster free memory is 56.94 GB which is coming from the node size 64GB. Even if I keep the value of h2o.external.memory to 24 GB, the H2O cluster free memory shows 56.94 GB. Shouldn't it show 24*2=48 GB?

<b>If I increase h2o.external.memory to more than 31 GB, then the cluster creation fails because the size of one single node is 64GB. In reality, we have 10 nodes, which sums up to 640 GB. But, external H2O is not using that memory. It's relying only on one node. How to make sure that H2O makes use of full memory available and scale according to the data.</b>

In addition to the above, if I try to increase cluster.size>2, it fails to create an h2o cluster.

```
Connecting to H2O server at http://h2o-mra-30m-exp6-driver-svc.md-ds.svc:54321 ... successful.
--------------------------  -------------------------------------------------------------------------------------------------------
H2O_cluster_uptime:         50 secs
H2O_cluster_timezone:       Etc/GMT
H2O_cluster_version:        3.32.1.3
H2O_cluster_name:           root
H2O_cluster_total_nodes:    2
H2O_cluster_free_memory:    56.94 Gb
H2O_cluster_total_cores:    2
H2O_cluster_allowed_cores:  2
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://h2o-mra-30m-exp6-8134487a3e82ef89-driver-svc.ua-ds.svc:54321
H2O_connection_proxy:       null
H2O_internal_security:      False
Python_version:             3.8.8 final
--------------------------  -------------------------------------------------------------------------------------------------------

Sparkling Water Context:
 * Sparkling Water Version: 3.32.1.3-1-3.1
 * H2O name: root
 * cluster size: 2
 * list of used nodes:
  (executorId, host, port)
  ------------------------
  (0,172.31.1.39,54321)
  (1,172.31.0.172,54321)
  ------------------------

  Open H2O Flow in browser: http://h2o-mra-30m-exp6-driver-svc.md-ds.svc:54321 (CMD + click in Mac OSX)
```

Logs: [logs-from-spark-kubernetes-driver-in-h2o-mra-30m-exp6-8134487a3e82ef89-driver for h2o github.txt](https://github.com/h2oai/sparkling-water/files/6710821/logs-from-spark-kubernetes-driver-in-h2o-mra-30m-exp6-8134487a3e82ef89-driver.for.h2o.github.txt)

Memory usage image: 

![memory_usage-30m-exp6](https://user-images.githubusercontent.com/23002214/123300660-6b4c9d00-d538-11eb-9c7c-e2ea8ba79577.PNG)

**Comment by Abhishekmamidi123** [comment]
@mn-mikke @jakubhava: I shared all the details and logs of the experiment above. Can you please tell us how to resolve the issue or what should be the correct configuration for the h2o backend to scale across nodes?

**Comment by mn-mikke** [comment]
Hi @Abhishekmamidi123,
I've finally got time to debug this issue and I've found out that it's a bug in Sparkling Water. The bug will be fixed in the next SW version `3.34.0.1-1` (Jira ticket: https://h2oai.atlassian.net/browse/SW-2581).

**Comment by mn-mikke** [comment]
Hi @Abhishekmamidi123, 
the bug fix has been already released in SW `3.32.1.5-1`. If the problem still persists, please reopen this ticket. I'm closing it for now.
